vortex-serde/src/layouts/pruning.rs

```diff
@@ -0,0 +1,538 @@
+    // TODO(robert): Add support for boolean column expressions,
+    //  i.e. if column is of bool dtype it's valid to filter on it directly as a predicate
+    if expr.as_any().downcast_ref::<Column>().is_some() {
+        return (fallback, HashMap::new());
+    }
+
+    if let Some(bexp) = expr.as_any().downcast_ref::<BinaryExpr>() {
+        if bexp.op() == Operator::Or || bexp.op() == Operator::And {
+            let (rewritten_left, mut refs_lhs) = convert_to_pruning_expression(bexp.lhs());
+            let (rewritten_right, refs_rhs) = convert_to_pruning_expression(bexp.rhs());
+            refs_lhs.extend(refs_rhs);
+            return (
+                Arc::new(BinaryExpr::new(rewritten_left, bexp.op(), rewritten_right)),
+                refs_lhs,
+            );
+        }
+
+        if let Some(col) = bexp.lhs().as_any().downcast_ref::<Column>() {
+            return PruningPredicateBuilder::try_new(col.field().clone(), bexp.op(), bexp.rhs())
+                .and_then(column_comparison_to_stat_expr)
+                .unwrap_or_else(|| (fallback, HashMap::new()));
+        };
+
+        if let Some(col) = bexp.rhs().as_any().downcast_ref::<Column>() {
+            return PruningPredicateBuilder::try_new(
+                col.field().clone(),
+                bexp.op().swap(),
+                bexp.lhs(),
+            )
+            .and_then(column_comparison_to_stat_expr)
+            .unwrap_or_else(|| (fallback, HashMap::new()));
+        };
+    }
+
+    (fallback, HashMap::new())
+}
+
+struct PruningPredicateBuilder<'a> {
+    column: Field,
+    operator: Operator,
+    other_exp: &'a Arc<dyn VortexExpr>,
+    references: HashMap<Field, ColumnStat>,
+}
+
+impl<'a> PruningPredicateBuilder<'a> {
+    pub fn try_new(
+        column: Field,
+        operator: Operator,
+        other_exp: &'a Arc<dyn VortexExpr>,
+    ) -> Option<Self> {
+        // TODO(robert): Simplify expression to guarantee that each column is not compared to itself
+        //  For majority of cases self column references are likely not prunable
+        if other_exp.references().contains(&column) {
+            return None;
+        }
+
+        Some(Self {
+            column,
+            operator,
+            other_exp,
+            references: HashMap::new(),
+        })
+    }
+
+    pub fn operator(&self) -> Operator {
+        self.operator
+    }
+
+    pub fn column_stat(&mut self, stat: Stat) -> Field {
+        let (new_field, stat_ref) = column_stat_reference(&self.column, stat);
+        self.references.insert(new_field.clone(), stat_ref);
+        new_field
+    }
+
+    pub fn rewrite_other_exp(&mut self, stat: Stat) -> Arc<dyn VortexExpr> {
+        replace_column_with_stat(self.other_exp, stat, &mut self.references)
+            .unwrap_or_else(|| self.other_exp.clone())
+    }
+
+    pub fn into_references(self) -> HashMap<Field, ColumnStat> {
+        self.references
+    }
+}
+
+fn column_comparison_to_stat_expr(
```

Comment:
WDYT about making the `build` or `rewrite` function of `PruningPredicateBuilder`?